Merge addTestCases to master

bananapizzza opened the pull request:
Add GetUserInput interface and GetUserInputWithScanner class.
Change from using Scanner to using GetUserInputWithScanner.
Add test cases for methods related to renting a locker.
Add test cases for methods related to opening a locker.

bananapizzza committed:
Add GetUserInput interface and GetUserInputWithScanner class.
Change from using Scanner to using GetUserInputWithScanner.
Add test cases for methods related to renting a locker.
Add test cases for methods related to opening a locker.

xaveng commented on src/main/java/CoinLocker/CoinLockerSystem.java:
```diff
@@ -47,18 +55,23 @@ public void startCoinLockerSystem() {
+                        openLocker();
+                        break;
+                    case 3:
+                        break;
```
Let's add a default case

bananapizzza replied:
added

xaveng commented on src/main/java/CoinLocker/CoinLockerSystem.java:
```diff
@@ -47,18 +55,23 @@ public void startCoinLockerSystem() {
+                        break;
+                }
+            } catch (InvalidInputTypeException e) {
+                System.out.println("Invalid input was entered. Please try again.");
```
Should we log error case to System.err ?

Maybe this could be refactored into into different log level later.

bananapizzza replied:
Changed to System.err

xaveng commented on src/main/java/CoinLocker/CoinLockerSystem.java:
```diff
@@ -154,7 +171,7 @@ void getMoney(int price) throws CancelRentingLockerException {
+            String choice = getUserInput.getString();
             switch (choice) {
                 case "#":
                     giveChange(0, paid);
```
I think in this case we should return the money, not the change.

Let's say the machine expects a user to insert $3 but the user only inserted $2 and press #, then we should return $2, not $-1.

xaveng replied:
Plus, I think giveChange should have a guard to check the price and paid before returning some value. Let's add testcases for these as well.

bananapizzza replied:
It works well

xaveng commented on src/main/java/CoinLocker/CoinLockerSystem.java:
```diff
@@ -215,13 +227,13 @@ private void openLocker() {
             System.out.println("Cancel opening a locker.");
         } catch (InvalidPasswordLimitReachedException e) {
-            System.out.println("You entered wrong password "+INVALID_PW_LIMIT+ " times. Please contact to the office.");
+            System.out.println("You entered wrong password " + INVALID_PW_LIMIT + " times. Please contact to the office.");
```
Should we mark the locker as permanently locked to prevent the user from retrying it again from the scratch?
If you thought about it but did not want to include it in this cr, TODO comment would be nice.

bananapizzza replied:
Added new list for temporarily locked lockers. Added logic and test cases related to the case.

xaveng commented on src/main/java/CoinLocker/GetUserInput.java:
```diff
@@ -0,0 +1,7 @@
+package CoinLocker;
+
+public interface GetUserInput {
```
For real project, we should add javadoc and comments for public interface and methods.

bananapizzza replied:
Javadoc added

xaveng commented on src/test/java/CoinLocker/CoinLockerSystemTest.java:
```diff
@@ -32,33 +39,53 @@ void test_addLockers() {
-
-            Assertions.assertThrows(NoAvailableLockerException.class, () -> coinLockerSystem.getAvailableLocker(LockerSize.SMALL));
+        } catch (InvalidLockerSizeException e) {
+            e.printStackTrace();
```
Are we expecting these exceptions will be thrown this test? If so we should use assertThrows, if not, we could just add exceptions to the method declaration and let JUnit handle those exceptions.

bananapizzza replied:
Changed

bananapizzza committed:
Add temporarilyLockedLockers for temporarily locked lockers.
Add logic and test cases related to the temporarily locked lockers.
Add Javadoc for GetUserInput interface.
Add exceptions to the test method declaration.
